analysis/src/analyzer.rs

```diff
@@ -9,38 +9,137 @@ use rustc_middle::mir;
+        where S: AbstractState<'a, 'tcx>
+    {
+        let mut p_state = PointwiseState::new(mir, self.tcx);
+        //use https://crates.io/crates/linked_hash_set for set preserving insertion order?
+        let mut work_set: BTreeSet<mir::BasicBlock> = BTreeSet::from_iter(mir.basic_blocks().indices());
 
-    pub fn definitely_initialized_analysis(
-        &self,
-        mir: &mir::Body<'tcx>,
-    ) -> Result<PointwiseState<DefinitelyInitializedState>> {
-        unimplemented!();
-    }
+        let mut counters: HashMap<mir::BasicBlock, u32> = HashMap::with_capacity(mir.basic_blocks().len());
+
+        'block_loop:    // extract the bb with the minimal index -> hopefully better performance
+        while let Some(&bb) = work_set.iter().next() {      //use pop_first when it becomes stable?
+            work_set.remove(&bb);
+
+            let mut state_before_block;
+            if bb == mir::START_BLOCK {
+                // entry block
+                state_before_block = S::new_initial(mir, self.tcx);
+            }
+            else {
+                state_before_block = S::new_bottom(mir, self.tcx);
+            }
+
+            for pred_bb in &mir.predecessors()[bb] {
+                if let Some(map) = p_state.lookup_after_block(pred_bb) {
+                    state_before_block.join(map.get(&bb).unwrap());      //map should contain bb, because we ensure that we have a state for every successor
+                }
+                // if no state is present: assume bottom => no effect on join
+            }
+
+            // widen if needed
+            let counter = counters.entry(bb).or_insert(0);
+            *counter += 1;
+
+            if S::need_to_widen(counter) {
+                let location = mir::Location {
+                    block: bb,
+                    statement_index: 0,
+                };
+                state_before_block.widen(p_state.lookup_before(&location).unwrap())
+            }
+
+            let mir::BasicBlockData { ref statements, .. } = mir[bb];
+            let mut current_state = state_before_block;
+            for statement_index in 0..statements.len() {
+                let location = mir::Location {
+                    block: bb,
+                    statement_index,
+                };
+                /* Too much performance overhead to check for every statement?
+                let prev_state = p_state.lookup_before(&location);
+                if prev_state.into_iter().any(|s| s == &current_state) {      //use .contains when it becomes stable
+                    // same state, don't need to reiterate
+                    continue 'block_loop;
+                }*/
+                p_state.set_before(&location, current_state.clone());
+                // normal statement
+                let res = current_state.apply_statement_effect(&location);
+                if res.is_err() {
+                    return Err(res.unwrap_err());
+                }
```

Comment:
Lines 91-94 can be replaced with `current_state.apply_statement_effect(&location)?;`. The [`?` operator](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html) makes the code look nicer.